### ISSUE TITLE
Infinite loop when trying to add an object to Realm at Watch Series 4 Simulator

### ISSUE THREAD

**Issue by ggirotto** [realm/realm-cocoa#5932]
## Goals
Add an object that inherits from Object to Realm

## Expected Results
The object is succesfull added

## Actual Results
Realms enters into an infinite loop and crashes with `EXC_BAD_ACCESS`. The loop has the following trace of functions 
`RLMAddObjectToRealm` at Realm.swift -> 
`Object::create` at object_accessor.hpp -> 
`Object::set_property_value_impl` at object_accessor.hpp -> 
`List::assign` at list.hpp -> 
`enumerate_list` at RLMAccessor.hpp ->
`List::add` at list.hpp ->
`List::dispatch(Fn&& fn)` at list.hpp ->
`switch_on_type` at property.hpp ->
`List::add` at list.hpp ->
`realm::RowExpr RLMAccessorContext::unbox` at RLMAccessor.mm ->
Finally again to `RLMAddObjectToRealm`

Below is the full stacktrace

```
* thread #1, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=1, address=0x4)
    frame #0: 0x0158c0a3 libobjc.A.dylib`objc_retain + 19
  * frame #1: 0x00698283 Realm`RLMAccessorContext::propertyValue(this=0xbff81000, obj=0x7d93ddc0, propIndex=6, prop=0x7d83a230) at RLMAccessor.mm:628 [opt]
    frame #2: 0x006990b9 Realm`RLMAccessorContext::value_for_property(this=0xbff81000, obj=0x7d93ddc0, (null)="customGroupedBy", propIndex=6) at RLMAccessor.mm:776 [opt]
    frame #3: 0x006bf320 Realm`realm::Object realm::Object::create<objc_object* __strong, RLMAccessorContext>(ctx=0xbff81000, realm=std::__1::shared_ptr<realm::Realm>::element_type @ 0x7d881750 strong=12 weak=5, object_schema=0x7aae9230, value=0x7d93ddc0, try_update=<unavailable>, out_row=0x7d93ddc4) at object_accessor.hpp:267 [opt]
    frame #4: 0x006bee22 Realm`::RLMAddObjectToRealm(object=0x7d93ddc0, realm=0x7d880a60, createOrUpdate=<unavailable>) at RLMObjectStore.mm:143 [opt]
    frame #5: 0x00698dbe Realm`realm::BasicRowExpr<realm::Table> RLMAccessorContext::unbox<realm::BasicRowExpr<realm::Table> >(this=0xbff810e8, v=0x7d93ddc0, create=<unavailable>, update=<unavailable>) at RLMAccessor.mm:747 [opt]
    frame #6: 0x0069ea41 Realm`void realm::Object::set_property_value_impl<objc_object* __strong, RLMAccessorContext>(this=0xbff81208, ctx=0xbff81230, property=0x7b73ea40, value=<unavailable>, try_update=<unavailable>, is_default=<unavailable>) at object_accessor.hpp:101 [opt]
    frame #7: 0x006bf41e Realm`realm::Object realm::Object::create<objc_object* __strong, RLMAccessorContext>(ctx=0xbff81230, realm=std::__1::shared_ptr<realm::Realm>::element_type @ 0x7d881750 strong=12 weak=5, object_schema=0x7aae9260, value=0x7d93e710, try_update=<unavailable>, out_row=0x7d93e714) at object_accessor.hpp:281 [opt]
    frame #8: 0x006bee22 Realm`::RLMAddObjectToRealm(object=0x7d93e710, realm=0x7d880a60, createOrUpdate=<unavailable>) at RLMObjectStore.mm:143 [opt]
    frame #9: 0x00698dbe Realm`realm::BasicRowExpr<realm::Table> RLMAccessorContext::unbox<realm::BasicRowExpr<realm::Table> >(this=0xbff81458, v=0x7d93e710, create=<unavailable>, update=<unavailable>) at RLMAccessor.mm:747 [opt]
    frame #10: 0x006c1039 Realm`auto realm::List::dispatch<void realm::List::add<objc_object* __strong&, RLMAccessorContext>(RLMAccessorContext&, objc_object* __strong&&&, bool)::'lambda'(objc_object* __strong&)>(objc_object* __strong&&&) const [inlined] auto void realm::List::add<objc_object* __strong&, RLMAccessorContext>(this=<unavailable>)::'lambda'(objc_object* __strong&)::operator()<realm::BasicRowExpr<realm::Table>*>(objc_object* __strong&) const at list.hpp:192 [opt]
    frame #11: 0x006c1023 Realm`auto realm::List::dispatch<void realm::List::add<objc_object* __strong&, RLMAccessorContext>(RLMAccessorContext&, objc_object* __strong&&&, bool)::'lambda'(objc_object* __strong&)>(objc_object* __strong&&&) const [inlined] auto realm::switch_on_type<void realm::List::add<objc_object* __strong&, RLMAccessorContext>(RLMAccessorContext&, objc_object* __strong&&&, bool)::'lambda'(objc_object* __strong&)>(type=<unavailable>, fn=0xbff81330) at property.hpp:171 [opt]
    frame #12: 0x006c0e7d Realm`auto realm::List::dispatch<void realm::List::add<objc_object* __strong&, RLMAccessorContext>(RLMAccessorContext&, objc_object* __strong&&&, bool)::'lambda'(objc_object* __strong&)>(this=0xbff81410, fn=0xbff81330) const at list.hpp:174 [opt]
    frame #13: 0x0069ecd4 Realm`void RLMAccessorContext::enumerate_list<void realm::List::assign<objc_object* __strong&, RLMAccessorContext>(RLMAccessorContext&, objc_object* __strong&&&, bool)::'lambda'(objc_object* __strong&&&)>(objc_object*, objc_object* __strong&&&) [inlined] void realm::List::add<objc_object* __strong&, RLMAccessorContext>(this=<unavailable>, ctx=<unavailable>, value=0x7d93e710) at list.hpp:192 [opt]
    frame #14: 0x0069ecb0 Realm`void RLMAccessorContext::enumerate_list<void realm::List::assign<objc_object* __strong&, RLMAccessorContext>(RLMAccessorContext&, objc_object* __strong&&&, bool)::'lambda'(objc_object* __strong&&&)>(objc_object*, objc_object* __strong&&&) [inlined] auto void realm::List::assign<objc_object* __strong&, RLMAccessorContext>(this=<unavailable>, element=0x7d93e710)::'lambda'(objc_object* __strong&&&)::operator()<objc_object* __strong&>(objc_object* __strong&&&) const at list.hpp:218 [opt]
    frame #15: 0x0069eca8 Realm`void RLMAccessorContext::enumerate_list<void realm::List::assign<objc_object* __strong&, RLMAccessorContext>(RLMAccessorContext&, objc_object* __strong&&&, bool)::'lambda'(objc_object* __strong&&&)>(this=0xbff81458, v=0x7da57110, func=0xbff813c8) at RLMAccessor.hpp:76 [opt]
    frame #16: 0x006c0e49 Realm`void realm::List::assign<objc_object* __strong&, RLMAccessorContext>(this=0xbff81410, ctx=0xbff81458, values=0xbff81498, update=<unavailable>) at list.hpp:217 [opt]
    frame #17: 0x0069e7f1 Realm`void realm::Object::set_property_value_impl<objc_object* __strong, RLMAccessorContext>(this=0xbff81578, ctx=0xbff815a0, property=0x7b73dd70, value=0x7da57110, try_update=<unavailable>, is_default=<unavailable>) at object_accessor.hpp:93 [opt]
    frame #18: 0x006bf41e Realm`realm::Object realm::Object::create<objc_object* __strong, RLMAccessorContext>(ctx=0xbff815a0, realm=std::__1::shared_ptr<realm::Realm>::element_type @ 0x7d881750 strong=12 weak=5, object_schema=0x7aae93b0, value=0x7da572d0, try_update=<unavailable>, out_row=0x7da572d4) at object_accessor.hpp:281 [opt]
    frame #19: 0x006bee22 Realm`::RLMAddObjectToRealm(object=0x7da572d0, realm=0x7d880a60, createOrUpdate=<unavailable>) at RLMObjectStore.mm:143 [opt]
    frame #20: 0x00698dbe Realm`realm::BasicRowExpr<realm::Table> RLMAccessorContext::unbox<realm::BasicRowExpr<realm::Table> >(this=0xbff81688, v=0x7da572d0, create=<unavailable>, update=<unavailable>) at RLMAccessor.mm:747 [opt]
    frame #21: 0x0069ea41 Realm`void realm::Object::set_property_value_impl<objc_object* __strong, RLMAccessorContext>(this=0xbff817a8, ctx=0xbff817d0, property=0x7b73e3f0, value=<unavailable>, try_update=<unavailable>, is_default=<unavailable>) at object_accessor.hpp:101 [opt]
    frame #22: 0x006bf41e Realm`realm::Object realm::Object::create<objc_object* __strong, RLMAccessorContext>(ctx=0xbff817d0, realm=std::__1::shared_ptr<realm::Realm>::element_type @ 0x7d881750 strong=12 weak=5, object_schema=0x7aae9320, value=0x7da57a60, try_update=<unavailable>, out_row=0x7da57a64) at object_accessor.hpp:281 [opt]
    frame #23: 0x006bee22 Realm`::RLMAddObjectToRealm(object=0x7da57a60, realm=0x7d880a60, createOrUpdate=<unavailable>) at RLMObjectStore.mm:143 [opt]
    frame #24: 0x00698dbe Realm`realm::BasicRowExpr<realm::Table> RLMAccessorContext::unbox<realm::BasicRowExpr<realm::Table> >(this=0xbff819f8, v=0x7da57a60, create=<unavailable>, update=<unavailable>) at RLMAccessor.mm:747 [opt]
    frame #25: 0x006c1039 Realm`auto realm::List::dispatch<void realm::List::add<objc_object* __strong&, RLMAccessorContext>(RLMAccessorContext&, objc_object* __strong&&&, bool)::'lambda'(objc_object* __strong&)>(objc_object* __strong&&&) const [inlined] auto void realm::List::add<objc_object* __strong&, RLMAccessorContext>(this=<unavailable>)::'lambda'(objc_object* __strong&)::operator()<realm::BasicRowExpr<realm::Table>*>(objc_object* __strong&) const at list.hpp:192 [opt]
    frame #26: 0x006c1023 Realm`auto realm::List::dispatch<void realm::List::add<objc_object* __strong&, RLMAccessorContext>(RLMAccessorContext&, objc_object* __strong&&&, bool)::'lambda'(objc_object* __strong&)>(objc_object* __strong&&&) const [inlined] auto realm::switch_on_type<void realm::List::add<objc_object* __strong&, RLMAccessorContext>(RLMAccessorContext&, objc_object* __strong&&&, bool)::'lambda'(objc_object* __strong&)>(type=<unavailable>, fn=0xbff818d0) at property.hpp:171 [opt]
    frame #27: 0x006c0e7d Realm`auto realm::List::dispatch<void realm::List::add<objc_object* __strong&, RLMAccessorContext>(RLMAccessorContext&, objc_object* __strong&&&, bool)::'lambda'(objc_object* __strong&)>(this=0xbff819b0, fn=0xbff818d0) const at list.hpp:174 [opt]
    frame #28: 0x0069ecd4 Realm`void RLMAccessorContext::enumerate_list<void realm::List::assign<objc_object* __strong&, RLMAccessorContext>(RLMAccessorContext&, objc_object* __strong&&&, bool)::'lambda'(objc_object* __strong&&&)>(objc_object*, objc_object* __strong&&&) [inlined] void realm::List::add<objc_object* __strong&, RLMAccessorContext>(this=<unavailable>, ctx=<unavailable>, value=0x7da57a60) at list.hpp:192 [opt]
    frame #29: 0x0069ecb0 Realm`void RLMAccessorContext::enumerate_list<void realm::List::assign<objc_object* __strong&, RLMAccessorContext>(RLMAccessorContext&, objc_object* __strong&&&, bool)::'lambda'(objc_object* __strong&&&)>(objc_object*, objc_object* __strong&&&) [inlined] auto void realm::List::assign<objc_object* __strong&, RLMAccessorContext>(this=<unavailable>, element=0x7da57a60)::'lambda'(objc_object* __strong&&&)::operator()<objc_object* __strong&>(objc_object* __strong&&&) const at list.hpp:218 [opt]
    frame #30: 0x0069eca8 Realm`void RLMAccessorContext::enumerate_list<void realm::List::assign<objc_object* __strong&, RLMAccessorContext>(RLMAccessorContext&, objc_object* __strong&&&, bool)::'lambda'(objc_object* __strong&&&)>(this=0xbff819f8, v=0x7d93f860, func=0xbff81968) at RLMAccessor.hpp:76 [opt]
    frame #31: 0x006c0e49 Realm`void realm::List::assign<objc_object* __strong&, RLMAccessorContext>(this=0xbff819b0, ctx=0xbff819f8, values=0xbff81a38, update=<unavailable>) at list.hpp:217 [opt]
    frame #32: 0x0069e7f1 Realm`void realm::Object::set_property_value_impl<objc_object* __strong, RLMAccessorContext>(this=0xbff81b18, ctx=0xbff81b40, property=0x7b73e020, value=0x7d93f860, try_update=<unavailable>, is_default=<unavailable>) at object_accessor.hpp:93 [opt]
    frame #33: 0x006bf41e Realm`realm::Object realm::Object::create<objc_object* __strong, RLMAccessorContext>(ctx=0xbff81b40, realm=std::__1::shared_ptr<realm::Realm>::element_type @ 0x7d881750 strong=12 weak=5, object_schema=0x7aae9380, value=0x7d93fd80, try_update=<unavailable>, out_row=0x7d93fd84) at object_accessor.hpp:281 [opt]
    frame #34: 0x006bee22 Realm`::RLMAddObjectToRealm(object=0x7d93fd80, realm=0x7d880a60, createOrUpdate=<unavailable>) at RLMObjectStore.mm:143 [opt]
    frame #35: 0x00dfbc6d RealmSwift`Realm.add(object=0x7d93fd80, update=<unavailable>, self=0x7b00ace0) at Realm.swift:287 [opt]
    frame #36: 0x000cdb54 Hercules-watchOS Extension`protocol witness for LocalDatabaseProtocol.add(_:update:) in conformance Realm at <compiler-generated>:0
    frame #37: 0x000cef15 Hercules-watchOS Extension`closure #1 in LocalPersistenceProtocol.saveLocally(self=0x7b150740, object=0x7d93fd80) at LocalPersistenceProtocol.swift:34
    frame #38: 0x000cef9c Hercules-watchOS Extension`partial apply for closure #1 in LocalPersistenceProtocol.saveLocally(_:) at <compiler-generated>:0
    frame #39: 0x000d0728 Hercules-watchOS Extension`closure #1 in LocalPersistenceProtocol.updateManagedObject(block=0x000cef70 Hercules-watchOS Extension`partial apply forwarder for closure #1 () -> () in (extension in Hercules_watchOS_Extension):Hercules_watchOS_Extension.LocalPersistenceProtocol.saveLocally(RealmSwift.Object) -> () at <compiler-generated>) at LocalPersistenceProtocol.swift:91
    frame #40: 0x000d077b Hercules-watchOS Extension`partial apply for closure #1 in LocalPersistenceProtocol.updateManagedObject(_:) at <compiler-generated>:0
    frame #41: 0x00dfb6a7 RealmSwift`Realm.write(block=0x000d0760 Hercules-watchOS Extension`partial apply forwarder for closure #1 () -> () in (extension in Hercules_watchOS_Extension):Hercules_watchOS_Extension.LocalPersistenceProtocol.updateManagedObject(() -> ()) -> () at <compiler-generated>, self=0x7b00ace0) at Realm.swift:155 [opt]
    frame #42: 0x000cdae8 Hercules-watchOS Extension`Realm.safeWrite(block=0x000d0760 Hercules-watchOS Extension`partial apply forwarder for closure #1 () -> () in (extension in Hercules_watchOS_Extension):Hercules_watchOS_Extension.LocalPersistenceProtocol.updateManagedObject(() -> ()) -> () at <compiler-generated>, self=0x7b00ace0, $error=(error_instance = Builtin.RawPointer)) at LocalDatabaseProtocol.swift:31
    frame #43: 0x000cdbfc Hercules-watchOS Extension`protocol witness for LocalDatabaseProtocol.safeWrite(_:) in conformance Realm at <compiler-generated>:0
    frame #44: 0x000d067a Hercules-watchOS Extension`LocalPersistenceProtocol.updateManagedObject(block=0x000cef70 Hercules-watchOS Extension`partial apply forwarder for closure #1 () -> () in (extension in Hercules_watchOS_Extension):Hercules_watchOS_Extension.LocalPersistenceProtocol.saveLocally(RealmSwift.Object) -> () at <compiler-generated>, self=0x7b150740) at LocalPersistenceProtocol.swift:90
    frame #45: 0x000ce5d6 Hercules-watchOS Extension`protocol witness for LocalPersistenceProtocol.updateManagedObject(_:) in conformance Persistence at <compiler-generated>:0
    frame #46: 0x000ced8b Hercules-watchOS Extension`LocalPersistenceProtocol.saveLocally(object=0x7d93fd80, self=0x7b150740) at LocalPersistenceProtocol.swift:33
    frame #47: 0x000ce40f Hercules-watchOS Extension`protocol witness for LocalPersistenceProtocol.saveLocally(_:) in conformance Persistence at <compiler-generated>:0
    frame #48: 0x000d833c Hercules-watchOS Extension`SessionManager.didReceiveModelUpdate(model=Any @ 0xbff821c0, type=user, self=0x001d9d2c) at SessionManager.swift:155
    frame #49: 0x000d79e9 Hercules-watchOS Extension`closure #1 in SessionManager.handleReceivedMessage(message=modelChanged, self=0x001d9d2c) at SessionManager.swift:134
    frame #50: 0x000d7bec Hercules-watchOS Extension`partial apply for closure #1 in SessionManager.handleReceivedMessage(_:) at <compiler-generated>:0
    frame #51: 0x000d629b Hercules-watchOS Extension`thunk for @escaping @callee_guaranteed () -> () at <compiler-generated>:0
    frame #52: 0x054c87b0 libdispatch.dylib`_dispatch_call_block_and_release + 15
    frame #53: 0x054c9bb5 libdispatch.dylib`_dispatch_client_callout + 14
    frame #54: 0x054d5eda libdispatch.dylib`_dispatch_main_queue_callback_4CF + 1217
    frame #55: 0x01f7090e CoreFoundation`__CFRUNLOOP_IS_SERVICING_THE_MAIN_DISPATCH_QUEUE__ + 14
    frame #56: 0x01f6ab7a CoreFoundation`__CFRunLoopRun + 3322
    frame #57: 0x01f69b47 CoreFoundation`CFRunLoopRunSpecific + 599
    frame #58: 0x01f6af3a CoreFoundation`CFRunLoopRunInMode + 122
    frame #59: 0x08c1f65b GraphicsServices`GSEventRunModal + 71
    frame #60: 0x08c1f549 GraphicsServices`GSEventRun + 80
    frame #61: 0x0b946e68 UIKitCore`UIApplicationMain + 139
    frame #62: 0x0582683f libxpc.dylib`_xpc_objc_main + 528
    frame #63: 0x058291a8 libxpc.dylib`xpc_main + 158
    frame #64: 0x01180ad8 Foundation`-[NSXPCListener resume] + 183
    frame #65: 0x04ae1c88 PlugInKit`___lldb_unnamed_symbol439$$PlugInKit + 147
    frame #66: 0x005aadea WatchKit`main + 85
    frame #67: 0x05537a31 libdyld.dylib`start + 1
```

## Steps to Reproduce
I've just updated XCode to the latest release (10), as well with Realm (3.10.0) and tryied to add an object to Realm through Watch series 4 simulator.

## Version of Realm and Tooling
Latest Realm version 3.10.0
Xcode version: 10
iOS/OSX version: 12 & watchOS 5
Dependency manager + version: Carthage 0.29

**Comment by tgoyne** [comment]
That stack trace doesn't show an infinite loop. Adding an object which has a list of objects to a Realm recursively adds all of those objects to the Realm as well.

It's crashing while trying to read a RealmOptional property from one of the objects. We've gotten a few reports of issues with this in the watchOS 4 simulator, but I haven't been able to reproduce it and no one has supplied any code which encounters problems.

**Comment by ggirotto** [comment]
Solved at #5923
